File: orderapp/OrderApp.java
// Java Console App, CRUD on Product - ID, Name, Description, Quantity, Unit Price. Menu with List, Add and Exit. Handle empty list, input mismatch, invalid choice errors and exception.

import java.util.ArrayList;
import java.util.InputMismatchException;
import java.util.List;
import java.util.Scanner;

public class OrderApp {
    private static final Scanner scanner = new Scanner(System.in);
    private static final List<Product> products = new ArrayList<>();

    public static void main(String[] args) {
        while (true) {
            System.out.println("1. List Products");
            System.out.println("2. Add Product");
            System.out.println("3. Exit");
            System.out.print("Enter choice: ");
            try {
                int choice = scanner.nextInt();
                scanner.nextLine();
                switch (choice) {
                    case 1:
                        listProducts();
                        break;
                    case 2:
                        addProduct();
                        break;
                    case 3:
                        System.out.println("Exiting...");
                        return;
                    default:
                        System.out.println("Invalid choice, try again.");
                }
            } catch (InputMismatchException e) {
                System.out.println("Invalid input, please enter a number.");
                scanner.nextLine();
            }
        }
    }

    private static void listProducts() {
        if (products.isEmpty()) {
            System.out.println("No products available.");
        } else {
            System.out.println("ID\tName\tDescription\tQuantity\tUnit Price");
            for (Product product : products) {
                System.out.println(product);
            }
        }
    }

    private static void addProduct() {
        System.out.print("Enter ID: ");
        int id = scanner.nextInt();
        scanner.nextLine();
        System.out.print("Enter Name: ");
        String name = scanner.nextLine();
        System.out.print("Enter Description: ");
        String description = scanner.nextLine();
        System.out.print("Enter Quantity: ");
        int quantity;
        try {
            quantity = scanner.nextInt();
        } catch (InputMismatchException e) {
            System.out.println("Invalid input for Quantity, please enter a valid integer.");
            scanner.nextLine();
            return;
        }
        
        System.out.print("Enter Unit Price: ");
        double unitPrice;
        try {
            unitPrice = scanner.nextDouble();
        } catch (InputMismatchException e) {
            System.out.println("Invalid input for Unit Price, please enter a valid decimal number.");
            scanner.nextLine();
            return;
        }
        scanner.nextLine();
        products.add(new Product(id, name, description, quantity, unitPrice));
        System.out.println("Product added successfully.");
    }
}

class Product {
    private final int id;
    private final String name;
    private final String description;
    private final int quantity;
    private final double unitPrice;

    public Product(int id, String name, String description, int quantity, double unitPrice) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.quantity = quantity;
        this.unitPrice = unitPrice;
    }

    @Override
    public String toString() {
        return id + "\t" + name + "\t" + description + "\t" + quantity + "\t" + unitPrice;
    }
}

File: orderapp/OrderApp.kt
import java.util.Scanner

fun main() {
    val scanner = Scanner(System.`in`)
    val products = mutableListOf<Product>()

    while (true) {
        /**
         * Prints the message "1. List Products" to the console.
         */
        println("1. List Products")
        println("2. Add Product")
        println("3. Exit")
        print("Enter choice: ")
        try {
            val choice = scanner.nextInt()
            scanner.nextLine() // Consume newline
            when (choice) {
                1 -> listProducts(products)
                2 -> addProduct(scanner, products)
                3 -> {
                    println("Exiting...")
                    return
                }
                else -> println("Invalid choice, try again.")
            }
        } catch (e: InputMismatchException) {
            println("Invalid input, please enter a number.")
            scanner.nextLine() // Consume the invalid input
        }
    }
}

fun listProducts(products: MutableList<Product>) {
    if (products.isEmpty()) {
        println("No products available.")
    } else {
        println("ID\tName\tDescription\tQuantity\tUnit Price")
        for (product in products) {
            println(product)
        }
    }
}

fun addProduct(scanner: Scanner, products: MutableList<Product>) {
    print("Enter ID: ")
    val id = scanner.nextInt()
    scanner.nextLine() // Consume newline
    print("Enter Name: ")
    val name = scanner.nextLine()
    print("Enter Description: ")
    val description = scanner.nextLine()
    print("Enter Quantity: ")
    val quantity = try {
        scanner.nextInt()
    } catch (e: InputMismatchException) {
        println("Invalid input for Quantity, please enter a valid integer.")
        scanner.nextLine() // Consume the invalid input
        return
    }

    print("Enter Unit Price: ")
    val unitPrice = try {
        scanner.nextDouble()
    } catch (e: InputMismatchException) {
        println("Invalid input for Unit Price, please enter a valid decimal number.")
        scanner.nextLine() // Consume the invalid input
        return
    }
    scanner.nextLine() // Consume newline
    products.add(Product(id, name, description, quantity, unitPrice))
    println("Product added successfully.")
}

data class Product(
    val id: Int,
    val name: String,
    val description: String,
    val quantity: Int,
    val unitPrice: Double
)